tools/ioflow/ioflow-common.h

```diff
@@ -0,0 +1,241 @@
+    val->offset = file->f_pos;
+
+    struct dentry *de = file->f_path.dentry;
+    bpf_probe_read_kernel(&val->file_name, sizeof(val->file_name), &de->d_iname);
+}
+
+static inline void comm_block_entry(struct val_t *val, u64 ts) {
+    val->ts_blk = ts;
+    if (val->ts_blk_start == 0)
+        val->ts_blk_start = ts;
+}
+
+static inline void comm_block_return(struct val_t *val, u64 ts) {
+    val->blk += ts - val->ts_blk;
+    val->ts_blk_end = ts;
+}
+
+static inline void comm_split_entry(struct val_t *val, u64 ts) {
+    val->cnt_split++;
+    val->ts_split = ts;
+    if (val->ts_split_start == 0)
+        val->ts_split_start = ts;
+}
+
+static inline void comm_split_return(struct val_t *val, u64 ts) {
+    val->split += ts - val->ts_split;
+    val->ts_split_end = ts;
+}
+
+static inline void comm_merge_entry(struct val_t *val, u64 ts) {
+    val->cnt_merge++;
+    val->ts_merge = ts;
+    if (val->ts_merge_start == 0)
+        val->ts_merge_start = ts;
+}
+
+static inline void comm_merge_return(struct val_t *val, u64 ts) {
+    val->merge += ts - val->ts_merge;
+    val->ts_merge_end = ts;
+}
+
+static inline void comm_rq_create(struct rqval_t *rqval, struct val_t *val, 
+                                  struct request *rq, u32 pid, u64 ts) {
+    val->cnt_blk++;
+    
+    rqval->pid = pid;
+    rqval->seq_num = val->seq_num;
+    rqval->ts_vfs = val->ts_vfs;
+    rqval->ts_rqcreate = ts;
+    rqval->sector = rq->__sector;
+    rqval->len = rq->__data_len;
+    struct gendisk *disk = rq->rq_disk;
+    bpf_probe_read_kernel(&rqval->disk_name, sizeof(rqval->disk_name), 
+        disk->disk_name);
+}
+
+static inline void comm_rq_issue(struct rqval_t *rqval, u64 ts) {
+    if (rqval) {
+        rqval->ts_rqissue = ts;
+    }
+}
+
+static inline void comm_rq_done(struct rqdata_t *rqdata, struct rqval_t *rqval, u64 ts) {
+    rqdata->pid = rqval->pid;
+    rqdata->seq_num = rqval->seq_num;
+    rqdata->ts_create = rqval->ts_rqcreate - rqval->ts_vfs;
+    if (rqval->ts_rqissue) {
+        rqdata->queue = rqval->ts_rqissue - rqval->ts_rqcreate;
+        rqdata->service = ts - rqval->ts_rqissue;
+    }
+    else {
+        rqdata->queue = 0;
+        rqdata->service = ts - rqval->ts_rqcreate;
+    }
+    
+    rqdata->sector = rqval->sector;
+    rqdata->len = rqval->len;
+    bpf_probe_read_kernel(&rqdata->disk_name, sizeof(rqdata->disk_name), 
+        &rqval->disk_name);
+}
+
+static inline void comm_vfs_return(struct data_t *data,  struct val_t *val, u32 pid, 
+                                   u64 ts, ssize_t size) {
+    data->pid = pid;
+    data->seq_num = val->seq_num;
+    data->total = ts - val->ts_vfs;
+
+    data->ts_vfs = val->ts_vfs;
+    data->ts_blk_start = val->ts_blk_start;
+    data->ts_blk_end = val->ts_blk_end;
+    data->ts_split_start = val->ts_split_start;
+    data->ts_split_end = val->ts_split_end;
+    data->ts_merge_start = val->ts_merge_start;
+    data->ts_merge_end = val->ts_merge_end;
+    data->blk = val->blk;
```

Comment:
Would it help to just calculate teh blk/split/merge time in comm_vfs_return, that way we can shrink val_t a bit?